code/core/src/shared/status-store/index.ts

```diff
@@ -0,0 +1,227 @@
+  typeId: StatusTypeId;
+  storyId: StoryId;
+  title: string;
+  description: string;
+  data?: any;
+  sidebarContextMenu?: boolean;
+}
+
+export const UNIVERSAL_STATUS_STORE_OPTIONS: StoreOptions<StatusesByStoryIdAndTypeId> = {
+  id: 'storybook/status',
+  leader: true,
+  initialState: {},
+} as const;
+
+const StatusStoreEventType = {
+  SELECT: 'select',
+} as const;
+
+export type StatusStoreEvent = {
+  type: typeof StatusStoreEventType.SELECT;
+  payload: Status[];
+};
+
+export type StatusStore = {
+  getAll: () => StatusesByStoryIdAndTypeId;
+  set: (statuses: Status[]) => void;
+  onAllStatusChange: (
+    listener: (
+      statuses: StatusesByStoryIdAndTypeId,
+      previousStatuses: StatusesByStoryIdAndTypeId
+    ) => void
+  ) => () => void;
+  onSelect: (listener: (selectedStatuses: Status[]) => void) => () => void;
+  unset: (storyIds?: StoryId[]) => void;
+};
+type FullStatusStore = StatusStore & {
+  selectStatuses: (statuses: Status[]) => void;
+  typeId: undefined;
+};
+export type StatusStoreByTypeId = StatusStore & {
+  typeId: StatusTypeId;
+};
+
+export type StatusStoreEnvironment = 'server' | 'manager' | 'preview';
+
+export type UseStatusStore = (selector?: (statuses: StatusesByStoryIdAndTypeId) => any) => any;
+
+export function createStatusStore(params: {
+  universalStatusStore: UniversalStore<StatusesByStoryIdAndTypeId, StatusStoreEvent>;
+  useUniversalStore?: never;
+  environment: StatusStoreEnvironment;
+}): {
+  getStatusStoreByTypeId: (typeId: StatusTypeId) => StatusStoreByTypeId;
+  fullStatusStore: FullStatusStore;
+};
+export function createStatusStore(params: {
+  universalStatusStore: UniversalStore<StatusesByStoryIdAndTypeId, StatusStoreEvent>;
+  useUniversalStore: typeof managerUseUniversalStore;
+  environment: StatusStoreEnvironment;
+}): {
+  getStatusStoreByTypeId: (typeId: StatusTypeId) => StatusStoreByTypeId;
+  fullStatusStore: FullStatusStore;
+  useStatusStore: UseStatusStore;
+};
+export function createStatusStore({
+  universalStatusStore,
+  useUniversalStore,
+  environment,
+}: {
+  universalStatusStore: UniversalStore<StatusesByStoryIdAndTypeId, StatusStoreEvent>;
+  useUniversalStore?: typeof managerUseUniversalStore;
+  environment: StatusStoreEnvironment;
+}): {
+  getStatusStoreByTypeId: (typeId: StatusTypeId) => StatusStoreByTypeId;
+  fullStatusStore: FullStatusStore;
+  useStatusStore?: UseStatusStore;
+} {
+  const fullStatusStore: FullStatusStore = {
+    getAll() {
+      return universalStatusStore.getState();
+    },
+    set(statuses) {
+      universalStatusStore.setState((state) => {
+        // Create a new state object to merge with the current state
+        const newState = { ...state };
+
+        // Process each status and merge it into the appropriate storyId record
+        for (const status of statuses) {
+          const { storyId, typeId } = status;
+
+          newState[storyId] = { ...(newState[storyId] ?? {}), [typeId]: status };
```

Comment:
Its a bit of a nit, and probably won't matter, but you're changing the order of keys here. A cleaner solution is to shallow copy the object first, then set the key on the copy after.

Reply:
yes that was actually an explicit intend, I felt that it was the right trade off to have simpler code here, and not have stable order. I don't know why anyone would need the order to be stable.